### ISSUE TITLE
Thanks to the author's work, can the author provide the salient mask M used for training?

### ISSUE THREAD

**Issue by reader-zora** [runjia0124/CoCoNet#13]

**Comment by runjia0124** [comment]
Hello, the salient mask is cropped and shared as part of the "finetune_data.h5" by google drive. See [here](https://github.com/runjia0124/CoCoNet/blob/2cda80959951edc5043157b7371fe24186f7c301/models/train_tasks.py#L200) on how it is loaded.

**Comment by reader-zora** [comment]
> Hello, the salient mask is cropped and shared as part of the "finetune_data.h5" by google drive. See [here](https://github.com/runjia0124/CoCoNet/blob/2cda80959951edc5043157b7371fe24186f7c301/models/train_tasks.py#L200) on how it is loaded.

Firstly, thank you very much for your reply. I have downloaded the files you mentioned, but the .h files contain preprocessed image data with a fixed size of 64×64, which prevents me from changing the image size.

Additionally, when reproducing your code, I noticed that your paper mentions cropping 1410 image patches of size 64×64 from the source images during the training phase. However, the training.h5 file you provided contains 6016 image patches. Could you clarify how many image patches were used for training?

Lastly, and most importantly, when strictly following the requirements outlined in your paper to reproduce your code, the loss fails to converge. What could be causing this issue?

I look forward to your response!

**Comment by runjia0124** [comment]
> > Hello, the salient mask is cropped and shared as part of the "finetune_data.h5" by google drive. See [here](https://github.com/runjia0124/CoCoNet/blob/2cda80959951edc5043157b7371fe24186f7c301/models/train_tasks.py#L200) on how it is loaded.
> 
> Firstly, thank you very much for your reply. I have downloaded the files you mentioned, but the .h files contain preprocessed image data with a fixed size of 64×64, which prevents me from changing the image size.
> 
> Additionally, when reproducing your code, I noticed that your paper mentions cropping 1410 image patches of size 64×64 from the source images during the training phase. However, the training.h5 file you provided contains 6016 image patches. Could you clarify how many image patches were used for training?
> 
> Lastly, and most importantly, when strictly following the requirements outlined in your paper to reproduce your code, the loss fails to converge. What could be causing this issue?
> 
> I look forward to your response!

Hello, thanks for your feedback.

1. Unfortunately I don't have the complete masks at hand, since we want to make training easier by sharing the preprocessed data. You may use common saliency detection / semantic segmentation methods to generate these masks. In the paper we found different ways to generate the masks do not have a big difference. If you don't mind, please send an email, I will get back to you if I find the data. 

2. the released model was trained as described in the paper, 1410 images, training with more patches should not have a big difference. 

3. did the loss decrease during training, if so, how was the performance after 5 epochs? We also found the loss curve fluctuates after a huge decrease, you may need to manually check the fusion result to decide a timing for early stopping, usually before 10 epochs.

Let me know if you have more questions!

**Comment by reader-zora** [comment]
> > > Hello, the salient mask is cropped and shared as part of the "finetune_data.h5" by google drive. See [here](https://github.com/runjia0124/CoCoNet/blob/2cda80959951edc5043157b7371fe24186f7c301/models/train_tasks.py#L200) on how it is loaded.
> > 
> > 
> > Firstly, thank you very much for your reply. I have downloaded the files you mentioned, but the .h files contain preprocessed image data with a fixed size of 64×64, which prevents me from changing the image size.
> > Additionally, when reproducing your code, I noticed that your paper mentions cropping 1410 image patches of size 64×64 from the source images during the training phase. However, the training.h5 file you provided contains 6016 image patches. Could you clarify how many image patches were used for training?
> > Lastly, and most importantly, when strictly following the requirements outlined in your paper to reproduce your code, the loss fails to converge. What could be causing this issue?
> > I look forward to your response!
> 
> Hello, thanks for your feedback.
> 
> 1. Unfortunately I don't have the complete masks at hand, since we want to make training easier by sharing the preprocessed data. You may use common saliency detection / semantic segmentation methods to generate these masks. In the paper we found different ways to generate the masks do not have a big difference. If you don't mind, please send an email, I will get back to you if I find the data.
> 2. the released model was trained as described in the paper, 1410 images, training with more patches should not have a big difference.
> 3. did the loss decrease during training, if so, how was the performance after 5 epochs? We also found the loss curve fluctuates after a huge decrease, you may need to manually check the fusion result to decide a timing for early stopping, usually before 10 epochs.
> 
> Let me know if you have more questions!


Thank you very much for your detailed response. My email is ng_oliver@outlook.com, and I look forward to your email.

    1. Regarding the first point, I will address the mask issue later.

    2.Regarding the second point, this question actually arises from the third point. Essentially, I was unable to successfully train your code, which led me to suspect if it could be due to the number of patches.

    3.Regarding the third point, I only changed the data_path in the code provided on your GitHub, but the loss does not converge. It fluctuates between 4-5 in each epoch and continues this way until the end of 30 epochs.

  Initially, I suspected that the issue might be with the loss function or that the gradient wasn't being reset, but after checking, I found no problems. Currently, I am still unable to identify the issue.

  Whether it is the method of generating masks or the number of patches, if I could successfully run the code, I could try modifying these aspects to see the corresponding results. Unfortunately, I have not yet been able to debug it successfully.

I look forward to your response! Thank you very much!

**Comment by reader-zora** [comment]
Hello,

Your code has been perplexing me for quite some time.

I have meticulously reviewed your code and found no apparent issues. However, the loss value never converges, with the minimum value only reaching 4, and it fluctuates between 4 and 5.

I then modified your loss function by removing the adaptive weight update and setting the weights to [[0.8, 0.2], [0.8, 0.2]]. Even then, the loss still did not converge, with the minimum value being 2. Is this behavior normal? How did the training go on your end?

Lastly, the loss function mentioned in your paper never converges in my code. How can I reproduce the weights you provided in your code?

I eagerly await your response.

Best regards.

**Comment by runjia0124** [comment]
> Hello,
> 
> Your code has been perplexing me for quite some time.
> 
> I have meticulously reviewed your code and found no apparent issues. However, the loss value never converges, with the minimum value only reaching 4, and it fluctuates between 4 and 5.
> 
> I then modified your loss function by removing the adaptive weight update and setting the weights to [[0.8, 0.2], [0.8, 0.2]]. Even then, the loss still did not converge, with the minimum value being 2. Is this behavior normal? How did the training go on your end?
> 
> Lastly, the loss function mentioned in your paper never converges in my code. How can I reproduce the weights you provided in your code?
> 
> I eagerly await your response.
> 
> Best regards.

Hello,

thanks for your feedback. The adaptive weight only controls the balance of the two modalities, it does not affect the loss curve. As I mentioned before, the loss did fluctuate after a huge descent, the problem exists in our training as well. Thus in our setting, it's okay if the loss fails to converge since the fusion result looks good. We manually checked the fusion performance to decide when to stop training.
